src/hockeypuck/pghkp/storage.go

```diff
@@ -91,6 +91,157 @@ var drConstraintsSQL = []string{
+rfingerprint IS NULL OR rsubfp IS NULL OR 
+EXISTS (SELECT 1 FROM subkeys_checked WHERE subkeys_checked.rsubfp = subkeys_copyin.rsubfp)
+`,
+	// Single-copy of in-file Dups && not Dups in DB
+	// Enforce foreign key constraint by checking both keys_checked and keys_copyin (instead of keys)
+	// *** ctid field is PostgreSQL-specific; Oracle has ROWID equivalent field ***
+	`INSERT INTO subkeys_checked (rfingerprint, rsubfp) 
+SELECT rfingerprint, rsubfp FROM subkeys_copyin WHERE 
+ctid IN 
+   (SELECT ctid FROM 
+      (SELECT ctid, ROW_NUMBER() OVER (PARTITION BY rsubfp ORDER BY ctid) rsubfpEnum FROM subkeys_copyin) AS dupRsubfpTAB 
+      WHERE rsubfpEnum = 1) AND 
+NOT EXISTS (SELECT 1 FROM subkeys WHERE subkeys.rsubfp = subkeys_copyin.rsubfp) AND 
+( EXISTS (SELECT 1 FROM keys_checked WHERE keys_checked.rfingerprint = subkeys_copyin.rfingerprint) OR 
+  EXISTS (SELECT 1 FROM keys_copyin  WHERE keys_copyin.rfingerprint  = subkeys_copyin.rfingerprint) )
+`, // Avoid checking "EXISTS (SELECT 1 FROM keys WHERE keys.rfingerprint = subkeys_copyin.rfingerprint)"
+	// by checking in keys_copyin (despite no indexing): only dups (in-file or _in DB_) still in keys_copyin
+	// -----------------
+	`INSERT INTO keys (rfingerprint, doc, ctime, mtime, md5, keywords) 
+SELECT rfingerprint, doc, ctime, mtime, md5, keywords FROM keys_checked
+`,
+	`INSERT INTO subkeys (rfingerprint, rsubfp) 
+SELECT rfingerprint, rsubfp FROM subkeys_checked
+`,
+}
+
+var bulkStatsSQL = []string{
+	`SELECT COUNT (*) FROM keys_copyin WHERE 
+rfingerprint IS NULL OR doc IS NULL OR ctime IS NULL OR mtime IS NULL OR md5 IS NULL
+`,
+	`SELECT COUNT (*) FROM subkeys_copyin WHERE 
+rfingerprint IS NULL OR rsubfp IS NULL
+`,
+	`SELECT COUNT (*) FROM keys_copyin WHERE 
+( ( NOT EXISTS (SELECT 1 FROM keys_checked WHERE keys_checked.rfingerprint = keys_copyin.rfingerprint OR
+                                                 keys_checked.md5          = keys_copyin.md5) AND 
+    ctid IN 
+        (SELECT ctid FROM 
+          (SELECT ctid, ROW_NUMBER() OVER (PARTITION BY rfingerprint ORDER BY ctid) rfpEnum FROM keys_copyin) AS dupRfpTAB 
+          WHERE rfpEnum = 1) ) OR 
+  (ctid IN 
+     (SELECT ctid FROM 
+        (SELECT ctid, ROW_NUMBER() OVER (PARTITION BY rfingerprint) rfpEnum FROM keys_copyin) AS dupRfpTAB 
+        WHERE rfpEnum > 1)) ) AND 
+NOT EXISTS (SELECT 1 FROM subkeys_checked WHERE subkeys_checked.rfingerprint = keys_copyin.rfingerprint)
+`,
+	`SELECT COUNT (*) FROM keys_copyin WHERE 
+ctid IN 
+   (SELECT ctid FROM 
+      (SELECT ctid, ROW_NUMBER() OVER (PARTITION BY rfingerprint) rfpEnum FROM keys_copyin) AS dupRfpTAB 
+      WHERE rfpEnum > 1) AND 
+EXISTS (SELECT 1 FROM subkeys_checked WHERE subkeys_checked.rfingerprint = keys_copyin.rfingerprint)
+`,
+	`SELECT COUNT (*) FROM keys_checked
+`,
+	`SELECT COUNT (*) FROM subkeys_checked
+`,
+}
+
+var keys_copyin_temp_table_name = "keys_copyin"
+var subkeys_copyin_temp_table_name = "subkeys_copyin"
```

Comment:
These could be consts

Reply:
On second thought, it might be best to not declare these at all, and avoid dynamic SQL...

Reply:
If I understand correctly what dynamic SQL is, there is none here. The strings resulting from `fmt.Sprintf(..)` using these table-name strings, along with up to 5K tuples of "($<n>, $<n+1>, ...)" of rolling <n> (line 960), are then given to `Prepare(...)` (see lines 898, 904 and 879).

This is based on this: https://stackoverflow.com/questions/12486436/how-do-i-batch-sql-statements-with-package-database-sql

Reply:
Done that last thing too. All `const`s now.